app/src/main/java/com/pileproject/drive/execution/ExecutionCondition.java

```diff
@@ -17,48 +17,196 @@
 package com.pileproject.drive.execution;
 
 
-import android.content.Context;
-
-import com.pileproject.drive.R;
-import com.pileproject.drive.app.DriveApplication;
 import com.pileproject.drive.programming.visual.block.BlockBase;
+import com.pileproject.drive.programming.visual.block.repetition.RepetitionEndBlock;
+import com.pileproject.drive.programming.visual.block.repetition.WhileForeverBlock;
 
 import java.util.ArrayList;
-import java.util.HashMap;
-import java.util.Map;
+import java.util.Collections;
+import java.util.List;
 import java.util.Stack;
 
 /**
- * This class has the condition of execution
- *
- * @author <a href="mailto:tatsuyaw0c@gmail.com">Tatsuya Iwanari</a>
- * @version 1.0 7-July-2013
+ * A container class that has the condition of program execution.
  */
 public class ExecutionCondition {
-    public static final int TRUE = 1;
-    public static final int FALSE = 0;
-    public Stack<Integer> whileStack;
-    public Stack<Map<String, Integer>> ifStack;
-    public int beginningOfCurrentWhileLoop;
-    public int programCount;
-    public ArrayList<BlockBase> blocks;
+    private Stack<Integer> mWhileStack;
+    private Stack<SelectionResult> mIfStack;
+    private int mBeginningOfCurrentLoop;
+    private int mProgramCount;
+    private final List<BlockBase> mBlocks;
+
+    public class SelectionResult {
+        public final int index;
+        public final boolean result;
+
+        public SelectionResult(int index, boolean result) {
+            this.index = index;
+            this.result = result;
+        }
+    }
+
+    public ExecutionCondition(ArrayList<BlockBase> blocks) {
+        mBlocks = blocks;
+        Collections.unmodifiableList(mBlocks);
+
+        mWhileStack = new Stack<>();
+        mIfStack = new Stack<>();
+
+        mBeginningOfCurrentLoop = -1;
+        mProgramCount = 0;
+    }
+
+    /**
+     * Check the program count is over than the program size or not.
+     * @return finished (true) or not (false)
+     */
+    public boolean hasProgramFinished() {
+        return mProgramCount >= mBlocks.size();
+    }
+
+    /**
+     * Get the specified block with an index.
+     * @param index
+     * @return
+     * @throws IndexOutOfBoundsException
+     */
+    public BlockBase getBlock(int index) throws IndexOutOfBoundsException {
```

Comment:
Since users of this class should not have access to blocks by indexing, it's better to hide this method by adding a getter with a selection result instance.